sycl/doc/extensions/InvokeSIMD/InvokeSIMD.asciidoc

```diff
@@ -0,0 +1,365 @@
+namespace ext {
+namespace oneapi {
+
+namespace simd_abi {
+template <class T, int N>
+using native_fixed_size = /* implementation-defined */;
+} // namespace simd_abi
+
+template <class T, int N>
+using simd = std::experimental::simd<T, simd_abi::native_fixed_size<N>>;
+
+template <class T, int N>
+using simd_mask = std::experimental::simd_mask<T, simd_abi::native_fixed_size<N>>;
+
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+----
+
+We introduce a new `native_fixed_size` ABI which combines elements of the
+existing `native` and `fixed_size` ABIs defined in Parallelism TS2;
+implementations are expected to use the most efficient representation
+of a `simd<T, N>` type (e.g. using multiple registers rather than an array).
+We expect implementations of this ABI to build on Clang's `ext_vector_type`,
+rather than specifying the register types directly.
+
+The value of `N` for each `simd` argument to a SIMD function must be the same,
+and represents the sub-group size of the calling kernel.  If a function accepts
+`simd` arguments with multiple `N` values, it cannot be invoked from a SPMD
+kernel and the implementation must issue a diagnostic in this case.
+
+The example below shows a simple SIMD function that scales all elements of
+an 8-wide SIMD type `x` by a scalar value `n`.
+
+[source, c++]
+----
+sycl::ext::oneapi::simd<float, 8> scale(sycl::ext::oneapi::simd<float, 8> x, float n)
+{
+  return x * n;
+}
+----
+
+=== Invoking a SIMD Function
+
+The `invoke_simd` free function invokes a SIMD function using all work-items
+in a `sub_group`.  The `invoke_simd` interface marshals data between the
+SPMD context of the calling kernel and the SIMD context of the callee,
+converting arguments and return values between scalar and SIMD types as
+appropriate.
+
+[source, c++]
+----
+namespace sycl {
+namespace ext {
+namespace oneapi {
+
+template <typename Callable, typename...T>
+auto invoke_simd(sub_group sg, Callable&& f, T... args);
+
+} // namespace oneapi
+} // namespace ext
+} // namespace sycl
+----
+
+The `Callable` passed to `invoke_simd` must be a C++ callable that can be
+invoked with the same number of arguments specified in the `args` parameter
+pack. `Callable` may be a function object, a lambda,
+or a function pointer (if the device supports `SPV_INTEL_function_pointers`).
+`Callable` must be an immutable callable with the same type and state for all
+work-items in the sub-group, otherwise behavior is undefined.
+
+Each argument to the invoked `Callable` must be a scalar type, a SIMD type,
+or a `std::tuple` of these types. The length of each SIMD type must match the
+sub-group size of the kernel; calling a SIMD function with mismatched lengths
+is an error, and an implementation must issue a diagnostic.
+
+To enable overloading in cases where the `Callable` accepts either a scalar
+type or a SIMD type, we leverage the `sycl::ext::oneapi::uniform` wrapper
+type from the `SYCL_EXT_ONEAPI_UNIFORM` extension.
```

Comment:
Even we discussed this in email, I would like to state my opinion here as I'm still not convinced by our discussion.  I believe having this special sycl::ext::oneapi::uniform is not needed in general. The compiler can match several possible callees by matching their data type, taking the "underlying" data type from vectors, so f(int ) and f(simd<int>) will both match. And then generate the call to the variant selected by uniformity analysis.  That strategy is perfectly aligned with our function pointers.

Reply:
We don't want the compiler to do the matching, and we don't want the compiler to do any uniformity analysis.  The purpose of `uniform` is to give developers complete control over matching (using regular C++ overload rules), and to let developers assert uniformity in situations where the compiler can't possibly prove it.

Reply:
The first statement is a little unclear. Who then will do matching?.. The suggested way just simplifies that matching.  
I'm not against this simplification, but I really don't like a "fully autonomous" uniform specification that can appear anywhere in the program.  That can lead to very hard-detectible mistakes, when user thinks the value is uniform but really it's not. E.g. a function parameter. When a user creates a function and uses it himself it may be always uniform.  But that can change when some one else uses that function.  I'll leave a comment in the uniform specification.

Reply:
Everything is handled by the C++ overloading mechanism and the type system.  A developer cannot accidentally declare something to be `uniform` by calling a function -- they have to call the `uniform` constructor explicitly.

Here's an example that might help:

```c++
// Function with varying (T) and uniform (uniform<T>) overloads
void foo(int x);
void foo(uniform<int> x);

// Function with varying (T) argument only
void bar(int x);

// Function with uniform (uniform<T>) argument only
void baz(uniform<T> x);

int main()
{
    int x;                // x is assumed varying
    foo(x);               // calls foo(int)
    foo(uniform<int>(x)); // calls foo(uniform<int>)

    bar(x);               // calls bar(int)
    bar(uniform<int>(x)); // calls bar(int) -- uniform<int> converts implicitly to int

    baz(x);               // compiler error -- int does not convert implicitly to uniform<int>
    baz(uniform<int>(x)); // calls baz(int)
}

Reply:
I meant the code inside the function. E.g.
void foo(int x) {
  int z = uniform<int>(x);
}
User can think that foo is called always with uniform parameter and assert that.

Reply:
This extension doesn't introduce that problem -- it attempts to solve it!

Today, if a parameter to a SYCL function must be uniform, that must be expressed in the function's documentation.  There is no way to describe via a function's signature that a given parameter must be uniform.  So, it's already possible to write functions that rely on assumed uniformity, like this:
```c++
void foo(sub_group sg, int x) {
  if (x == 5) {
    int z = reduce_over_group(sg, x);
  } else {
    ...
  }
}
```
...if `x` is not uniform, then it is illegal to call `reduce_over_group` inside of the conditional.  The user can document that `foo` requires `x` to be uniform, but cannot enforce it.  If they want to support both uniform- and non-uniform data, the only way to do so is via a run-time check:
```c++
void foo(sub_group sg, int x) {
  if (all_of_group(sg, x == 5)) {
    int z = reduce_over_group(sg, x);
  } else {
    ...
  }
}
```

This extension allows for the uniformity requirement to be declared as part of the function signature.  I don't think people would write the code from your example; they'd write the below instead:
```c++
void foo(uniform<int> x) {
  int z = x;
}
```

Reply:
Creating a spec should suppose some implementation. I would like to know how you see the implementation for IA.

Reply:
And one more question. How is this aligned with our function pointers specification? As far as I remember, we had the same intent,  explicit uniform/linear/varying specification for formal and actual parameters. Do you remember why that was not accepted? Are we going to change that spec too?